app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,464 @@
+                :export_tags, :tag_filters, :custom_json
+
+  validates :package_mode, :package_type, presence: true, if: :activated?
+
+  validates_each :custom_json,
+    if: :allow_custom_json_validation? do |record, attr, value|
+      begin
+        name_re = /([A-Za-z0-9&_-]+\/[A-Za-z0-9&_-]+)/
+
+        if value.empty?
+          error = 'must be specified'
+        elsif (cjson = ActiveSupport::JSON.decode(value))
+          if cjson.empty?
+            error = 'must not be empty'
+          elsif !cjson['name']
+            error = 'must have a name key specified'
+          elsif cjson['name'].empty?
+            error = 'name key must not be empty'
+          elsif cjson['name'] != (name_re.match(cjson['name']) || [])[0]
+            error = 'name key must be formatted as "namespace/project"'
+          elsif !cjson['description']
+            error = 'must have a description key specified'
+          elsif cjson['description'].empty?
+            error = 'description key must not be empty'
+          end
+        end
+
+        if error
+          record.errors.add(attr, error)
+        end
+
+      rescue
+        record.errors.add(attr, 'must be a valid JSON string')
+      end
+    end
+
+  after_save :process_project
+
+  def allow_custom_json_validation?
+    activated? && package_mode == 'advanced'
+  end
+
+  def title
+    'Composer'
+  end
+
+  def description
+    'This project will be publicly listed as a composer package,
+but usage of private and internal repositories will still
+require authentication.'
+  end
+
+  def help
+    out = 'This project will be publicly listed as a composer package,
+but usage of private and internal repositories will still
+require authentication. '
+    if package_mode == 'default'
+      out += 'The package is exported using the project\'s composer.json. '\
+      'Additional settings are ignored.'
+    elsif package_mode == 'project'
+      out += 'The package is exported using the project\'s attributes '\
+      'The following settings are applied.'
+    elsif package_mode == 'advanced'
+      out += 'The package is exported using the custom json specified '\
+      'in the configuration.'
+    end
+    out
+  end
+
+  def to_param
+    'composer'
+  end
+
+  def fields
+    [
+      { type: 'fieldset', legend: 'Packaging:', fields:
+        [
+          { type: 'select',
+            name: 'package_mode',
+            choices:
+            [
+              ['Built-in: parse composer.json (default)', 'default'],
+              ['Project: generate from project attributes', 'attributes'],
+              ['Advanced: use custom JSON', 'advanced']
+            ],
+            default_choice: 'default'
+          },
+          { type: 'select',
+            name: 'package_type',
+            hint: 'Applicable only on project package mode.',
+            choices:
+            [
+              ['AGL Module', 'agl-module'],
```

Comment:
Well I cant really say how often this list can be changed. That project would probably extend support to additional package types over time but as I mentioned, this is something that can also be achieved using Packagin Mode: Advanced and using a custom.json with any package type the gitlab user wants to use.

Alternatively, we can have this field set as a text input, and users will have to input the package type manually thus not relying to that project at all.